app/src/org/commcare/views/widgets/GregorianDateWidget.java

```diff
@@ -0,0 +1,336 @@
+                        selectedDate.set(Calendar.YEAR, Integer.parseInt(contents));
+                    }
+                }
+            }
+        });
+    }
+
+    private void clearAll(){
+        dayTxt.setText("");
+        monthTxt.setText("");
+        yearTxt.setText("");
+        super.setFocus(getContext());
+    }
+
+    @Override
+    protected void inflateView(Context context){
+        LayoutInflater vi = (LayoutInflater)context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+        gregorianView = (LinearLayout) vi.inflate(R.layout.gregorian_date_widget, null);
+        addView(gregorianView);
+    }
+
+    @Override
+    protected void updateDateDisplay(long millisFromJavaEpoch) {
+        UniversalDate dateUniv = fromMillis(millisFromJavaEpoch);
+        monthArrayPointer = dateUniv.month - 1;
+        dayTxt.setText(String.format(Locale.getDefault(), DAYFORMAT, dateUniv.day));
+        monthTxt.setText(monthsArray[monthArrayPointer]);
+        yearTxt.setText(String.format(Locale.getDefault(), YEARFORMAT, dateUniv.year));
+        selectedDate.setTimeInMillis(millisFromJavaEpoch);
+        dayOfWeek.setText(selectedDate.getDisplayName(Calendar.DAY_OF_WEEK, Calendar.LONG, Locale.getDefault()));
+    }
+
+    @Override
+    protected long getCurrentMillis() {
+
+        checkEmptyText();
+
+        int day = Integer.parseInt(dayTxt.getText().toString());
+        int month = monthArrayPointer + 1;
+        int year = Integer.parseInt(yearTxt.getText().toString());
+
+        if(day > selectedDate.getActualMaximum(Calendar.DAY_OF_MONTH)){
+            dayTxt.setText(String.valueOf(selectedDate.getActualMaximum(Calendar.DAY_OF_MONTH)));
+            day = Integer.parseInt(dayTxt.getText().toString());
+        }
+
+        return toMillisFromJavaEpoch(year, month, day, millisOfDayOffset);
+    }
+
+    private void checkEmptyText() {
+        if(dayTxt.getText().toString().equals("")){
+            dayTxt.setText(String.valueOf(selectedDate.get(Calendar.DAY_OF_MONTH)));
+        }
+
+        if(monthTxt.getText().toString().equals("")){
+            monthTxt.setText(monthsArray[monthArrayPointer]);
+        }
+
+        if(yearTxt.getText().toString().equals("")){
+            yearTxt.setText(String.valueOf(selectedDate.get(Calendar.YEAR)));
+        }
+    }
+
+    @Override
+    protected void updateGregorianDateHelperDisplay(){}
+
+    @Override
+    protected UniversalDate decrementMonth(long millisFromJavaEpoch) {
+        DateTime dt = new DateTime(millisFromJavaEpoch).minusMonths(1);
+        return constructUniversalDate(dt);
+    }
+
+    @Override
+    protected UniversalDate decrementYear(long millisFromJavaEpoch) {
+        DateTime dt = new DateTime(millisFromJavaEpoch).minusYears(1);
+        return constructUniversalDate(dt);
+    }
+
+    @Override
+    protected UniversalDate fromMillis(long millisFromJavaEpoch) {
+        return constructUniversalDate(new DateTime(millisFromJavaEpoch));

```

Comment:
You should verify that the bug fixed in https://github.com/dimagi/commcare-odk/pull/1144 for the Nepali date widget doesn't occur here. Reproduction steps are in the description and in the test files. Let me know if you want to sit down and look into this.